% Possible reviewers that I know personally include:
% - Chris Fields
% - Denis Baurain
% - Daisie Huang
% In addition, there are people that appear to be users but that I don't know:
% - Hayward, Alexander 
% - Jern, Patric
% The general idea here is that these are people I haven't published with (yet).
% The BioPerl people's page provides a murderer's row of other candidates, though
% quite a few I already have published with, so there's a potential for COI.
% Here's the list: https://github.com/orgs/bioperl/people
% Also, all people on the previous publication, or people that have since 
% been added to the Acknowledgements in the README.md are obviously out.



\documentclass{bioinfo}
\copyrightyear{2017} \pubyear{2017}

\access{Advance Access Publication Date: Day Month Year}
\appnotes{Applications Note}

% Application Notes (up to 2 pages; this is approx. 1300 words or 1000 words plus one figure)

\begin{document}
\firstpage{1}

\subtitle{Phylogenetics}

\title[short Title]{The Bio::Phylo libraries for phylogenetic data analysis, version 2.0}
\author[Sample \textit{et~al}.]{Rutger A. Vos\,$^{\text{\sfb 1,2}*}$ and Hannes Hettling\,$^{\text{\sfb 1}}$}
\address{$^{\text{\sf 1}}$Research group \textit{Endless Forms}, Naturalis Biodiversity Center, Leiden, P.O. Box 9517, 2300RA, The Netherlands and \\
$^{\text{\sf 2}}$Institute of Biology Leiden, Leiden University, Leiden, P.O. Box 9500, 2300RA, The Netherlands.}

\corresp{$^\ast$To whom correspondence should be addressed.}

\history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}

\editor{Associate Editor: XXXXXXX}

\abstract{\textbf{Motivation:} Phylogenetic analysis is a broad and expanding field that
requires versatile programming toolkits to manage the various data types, file formats,
and needs for scalability, simulation, visualization, and data exploration.\\
\textbf{Results:} We present version 2.0 of the Bio::Phylo libraries for phylogenetic data
analysis. This new release represents a rewrite of the internal architecture, allowing
for extensions that improve speed and persistence, as well as greatly increased 
functionality in terms of analysis, data reading and writing, and visualization.\\
\textbf{Availability:} The package is released as open source software under the same terms as 
Perl itself and available from the comprehensive Perl archive network as well as directly
from the source code repository.\\
\textbf{Contact:} \href{rutger.vos@naturalis.nl}{rutger.vos@naturalis.nl}\\
\textbf{Supplementary information:} Supplementary data are available at \textit{Bioinformatics}
online.}

\maketitle

% RAV: here are some useful syntax tricks that we will need to observe:
% - refer to figures like so: 'Figure~\ref{fig:01}'
% - citations like this: '\citep{Bag01}' or like that: '\citet{Bag01}'
% - equations as follows:
% \begin{equation}
% \sum \text{\it x}+ \text{\it y} =\text{\it Z}\label{eq:01}\vspace*{-10pt}
% \end{equation}
% which can then be referenced as Equation~(\ref{eq:01})

\section{Introduction}

The analysis of phylogenetic trees and associated data is a common operation in 
bioinformatics. To do so in the face of large data volumes requires automation, for 
example using scripting toolkits. The Bio::Phylo libraries \citep{Vos2011} provide such 
functionality for the Perl programming language, in a modular toolkit whose components are 
used in anything from quick scripts, e.g. to process some data as part of a larger 
analysis (numerous publications cite this usage, some high-impact ones among these include 
\citet{Hayward2015} and \citet{Smet2013}), to larger applications, such as the SUPERSMART 
pipeline \citep{Antonelli2017} or the Monophylizer \citep{Mutanen2016} web service. In 
addition, other libraries for bioinformatics and computational biology depend on 
Bio::Phylo in order to re-use some of its functionality. The most widely-used and 
well-known of these is BioPerl \citep{Stajich2002}, which uses Bio::Phylo for reading and 
writing NeXML \citep{Vos2012} data. 

The Bio::Phylo libraries are designed around the notion that phylogenetic data can be 
modeled as consisting of the properties of, and relationships between, objects from a 
limited number of classes. These classes are things such as phylogenetic trees, the tree 
nodes in them, and the taxa that terminal tree nodes represent; or the characters of 
character state matrices, and the character state sequences in them. Bio::Phylo provides 
the functionality to create these objects from scratch, to read them from various data 
sources, or to simulate them; and, subsequently, to transform them, perform calculations 
on them, and to report the results of these operations by exporting data in various 
formats or by visualizing them. Some application areas can be identified for which 
Bio::Phylo is particularly well suited. These are briefly described in the following 
sections, with emphasis on new features implemented in version 2.0.

\section{Description}

The Bio::Phylo API models a number of relationships between different classes of data 
objects. For example, a phylogenetic tree has zero or more nodes in it, and, conversely, 
a node belongs at most to a single tree. These relationships, which can be termed "has-a" 
and "has-many" relationships, have been exhaustively described previously (see 
\citet{Vos2011} and Figure 1 therein) and have not changed in version 2.0. In addition, 
the API models relationships between classes of objects in terms of their inheritance from 
one another, i.e. "is-a" relationships. In the new API, these latter relationships have 
subtly changed, although this does not affect the behavior experienced by users of the 
toolkit. These changes are discussed in the section on internal design.

\begin{itemize}
\item Instantiating objects standalone, using factory, from files, using the new parsers: 
Adjacency, CDAO \citep{prosdocimi2009initial}, FASTQ, Figtree \citep{rambaut2007figtree}, 
NHX \citep{zmasek2001atv}
\item Exporting objects to text files using the new unparsers: Adjacency, CDAO 
\citep{prosdocimi2009initial}, Figtree \citep{rambaut2007figtree}, HTML, NHX 
\citep{zmasek2001atv}, Nwmsrdf
\end{itemize}

\subsection{NeXML management}

\begin{itemize}
\item NeXML description
\item Semantic annotations
\item Other representations of NeXML: RDF and JSON
\item Required dependencies
\end{itemize}

% Here's how to insert a table. We might want to use this for a brief overview of the
% new IO options (for example). Here's how to reference such a table: 'Table~\ref{Tab:01}'
\begin{table}[!t]
\processtable{This is table caption\label{Tab:01}} {\begin{tabular}{@{}llll@{}}\toprule head1 &
head2 & head3 & head4\\\midrule
row1 & row1 & row1 & row1\\
row2 & row2 & row2 & row2\\
row3 & row3 & row3 & row3\\
row4 & row4 & row4 & row4\\\botrule
\end{tabular}}{This is a footnote}
\end{table}

% Here's how to insert a figure. We'll probably use this for examples of visualization
\begin{figure}[!tpb]%figure1
\fboxsep=0pt\colorbox{gray}{\begin{minipage}[t]{235pt} \vbox to 100pt{\vfill\hbox to
235pt{\hfill\fontsize{24pt}{24pt}\selectfont FPO\hfill}\vfill}
\end{minipage}}
%\centerline{\includegraphics{fig01.eps}}
\caption{Caption, caption.}\label{fig:01}
\end{figure}

%\begin{figure}[!tpb]%figure2
%%\centerline{\includegraphics{fig02.eps}}
%\caption{Caption, caption.}\label{fig:02}
%\end{figure}

\subsection{Visualization}

\begin{itemize}
\item cladograms versus phylograms
\item projections: rectangular, unrooted, radial
\item annotations: collapsed clades, clade labels, pie charts
\item required dependencies
\end{itemize}

\subsection{Molecular evolution}

\begin{itemize}
\item substitution model objects
\item model testing
\item model serialization
\end{itemize}

\subsection{Simulation}

\begin{itemize}
\item basic tree simulation and evolutionary models
\item simulation of expected waiting times
\item simulating sequence data
\end{itemize}

\subsection{Web services}

\begin{itemize}
\item clients
\item services
\item PhyloWS
\item BioVeL
\item Monophylizer
\item NeXML website
\end{itemize}

\subsection{Compiled extensions}

\begin{itemize}
\item Bio::Phylo::XS
\item Beagle
\item PLL
\end{itemize}

\subsection{Database mappings}

\begin{itemize}
\item SUPERSMART
\item Bio::Phylo::Megatree
\end{itemize}

\subsection{Internal design}

\begin{itemize}
\item Semantic annotations
\item "Roles"
\item Meta-Object Programming
\end{itemize}

\section{Installation}

Bio::Phylo can be installed automatically by a number of different package managers. 
However, these usually rely on outdated versions, and because there is nothing that needs 
to be compiled for the core installation there is no need to let a package manager do a 
system wide installation, which is typically only useful for packages that include 
compiled binaries and not just pure Perl source code. Bio::Phylo is pure Perl, so it is 
recommended to have the project's directory tree "cloned" from the source code repository 
(see Availability), and have the \texttt{PERL5LIB} environment variable point to the 
\texttt{lib} folder inside the directory tree. This way all functionality will be 
available in user scripts, and the user can update to the latest release (or skip back to 
any previous one) with a single command. Source code can be cloned from code repositories
using the \texttt{git} (http://git-scm.org) tool on most common operating systems, it is 
outside of the scope of this document to discuss how this is done: good documentation for 
\texttt{git} can be found online (e.g. 
https://help.github.com/articles/cloning-a-repository/). How to update an environment 
variable to add an additional location to it, on any particular operating system, is 
likewise best searched for online.

Although this is not the recommended way to install, Bio::Phylo is also available from 
CPAN, which means that any tools built around this system will be able to install the 
latest release, for example using the standard command line program \texttt{cpan}, which 
is installed on almost any computer that has \texttt{perl} installed. Installation from 
CPAN is typically the approach taken by a system administrator on high-performance 
computing infrastructure. Experienced Perl users commonly install the non-standard package 
management tool \texttt{cpanm} (CPAN-minus, a play on the name of a previous tool called 
CPAN-plus), which is somewhat more user friendly, and it can install modules directly from 
GitHub (but then it will do a system-wide install that from that moment on cannot be 
updated from GitHub, so it is still better to check out the code as described above). 
Certain installations of Perl (historically, on Windows) include a package manager called 
\texttt{ppm}. While this is very useful for hard-to-install compiled extensions, this too 
relies on old versions of Bio::Phylo, so is best avoided, as are Linux-based package 
managers, such as \texttt{apt-get} and \texttt{yum}.

\section{Availability}

All revisions of the source code are available from the source code repository at
http://github.com/rvosa/bio-phylo. The latest stable release version, which is usually 
considerably far behind the latest source code revision, is available from the 
Comprehensive Perl Archive Network (CPAN) at http://search.cpan.org/dist/Bio-Phylo. 
Accompanying this publication is a uniquely identifiable release, stamped with a Digital 
Object Identifier (DOI) issued by Zenodo.org: \texttt{doi:10.5281/zenodo.tbd}.

As is a common convention in Perl software releases, a dual licensing scheme applies to
Bio::Phylo - both the Artistic License 
(https://github.com/rvosa/bio-phylo/blob/master/COPYING) as well as the GNU General Public 
License (https://github.com/rvosa/bio-phylo/blob/master/LICENSE) applies. This is 
generally interpreted to mean that you are free to choose whichever of these licenses fits 
best with your own project, should you want to reuse all (or part) of Bio::Phylo. This is 
certainly the spirit: feel free to use these libraries however you see fit. No warranties.

\section*{Acknowledgements}

\section*{Funding}

The research leading to these results has received funding from the European Community's 
Seventh Framework Programme (FP7/2007-2013) under grant agreement no. 237046.

\bibliographystyle{natbib}

\bibliography{document}

\end{document}
